### PR TITLE
[OTA-3739] Refactor login process.

### PR DESCRIPTION
**Pull request by houcros**
To be merged with https://github.com/advancedtelematic/ota-plus-user-profile/pull/106.

### CODE REVIEW

File: ota-plus-web/app/com/advancedtelematic/auth/oidc/NamespaceProvider.scala [hunk 1]
```diff
@@ -22,19 +24,33 @@ class NamespaceFromIdentity extends NamespaceProvider {
...
-        organizations.filter(_.isDefault).map(_.namespace).head
-    }
+    val userId = tokens.idToken.userId
+    userProfileApi
+      .getUser(userId)
+      .map(namespaceFromJson)
+      .recoverWith {
+        case RemoteApiError(result, _) if result.header.status == 404 =>
+          createUser(tokens.accessToken, userId)
+      }
```
Comment (koshelev) [inline ota-plus-web/app/com/advancedtelematic/auth/oidc/NamespaceProvider.scala]: I would suggest to move `.map(namespaceFromJson)` here, make it a last call

File: ota-plus-web/app/com/advancedtelematic/auth/oidc/NamespaceProvider.scala [hunk 2]
```diff
@@ -22,19 +24,33 @@ class NamespaceFromIdentity extends NamespaceProvider {
...
-        Namespace.generate
-      case organizations =>
-        organizations.filter(_.isDefault).map(_.namespace).head
-    }
+    val userId = tokens.idToken.userId
+    userProfileApi
+      .getUser(userId)
+      .map(namespaceFromJson)
+      .recoverWith {
+        case RemoteApiError(result, _) if result.header.status == 404 =>
```
Comment (koshelev) [inline ota-plus-web/app/com/advancedtelematic/auth/oidc/NamespaceProvider.scala]: Is there any chance user profile also returns an ErrorRepresentation and we can see that 404 is not a broken routing or something but a specific response for a non existing user?

File: ota-plus-web/app/com/advancedtelematic/auth/oidc/NamespaceProvider.scala (hunk 1 above)
Comment (houcros) [reply ota-plus-web/app/com/advancedtelematic/auth/oidc/NamespaceProvider.scala]: Done, amended in the same commit.

File: ota-plus-web/app/com/advancedtelematic/auth/oidc/NamespaceProvider.scala (hunk 2 above)
Comment (houcros) [reply ota-plus-web/app/com/advancedtelematic/auth/oidc/NamespaceProvider.scala]: I found a way to check for that but I'm not too happy with it, see my last commit. I don't think there is an easy way to check for a specific remote API error like that without doing some refactoring around our API clients...
So in my opinion I'd leave that out of this PR. How likely is that route would be broken? I'd venture not much...

Comment (houcros) [reply ota-plus-web/app/com/advancedtelematic/auth/oidc/NamespaceProvider.scala]: I'll merge this for now and open a new PR with that last commit if we want it :)

Comment (houcros) [reply ota-plus-web/app/com/advancedtelematic/auth/oidc/NamespaceProvider.scala]: @koshelev see this one for the commit I was talking about https://github.com/advancedtelematic/ota-plus-server/pull/2094.

**Pull request merged**